app/components/syntax-highlighted-diff.ts

```diff
@@ -161,27 +175,57 @@ export default class SyntaxHighlightedDiffComponent extends Component<Signature>
-      return; // Not ready to highlight yet
+    // Read all dependent parameters into local constants
+    const {
+      codeWithoutDiffMarkers,
+      args: { code, language },
+      darkMode: { isEnabled: isDarkMode },
+    } = this;
+
+    // Return if nothing changed since last render
+    if (
+      this.#lastHighlightCodeParams.compareWithParams({
+        isDarkMode: isDarkMode,
+        language: language,
+        codeWithoutDiffMarkers: codeWithoutDiffMarkers,
+      })
+    ) {
+      return;
     }
 
+    // Prepare the highlighter promise
     const highlighterPromise = getOrCreateCachedHighlighterPromise(
-      this.isDarkMode ? SyntaxHighlightedDiffComponent.highlighterIdForDarkMode : SyntaxHighlightedDiffComponent.highlighterIdForLightMode,
-      this.isDarkMode ? SyntaxHighlightedDiffComponent.highlighterOptionsForDarkMode : SyntaxHighlightedDiffComponent.highlighterOptionsForLightMode,
+      isDarkMode ? SyntaxHighlightedDiffComponent.highlighterIdForDarkMode : SyntaxHighlightedDiffComponent.highlighterIdForLightMode,
+      isDarkMode ? SyntaxHighlightedDiffComponent.highlighterOptionsForDarkMode : SyntaxHighlightedDiffComponent.highlighterOptionsForLightMode,
     );
 
-    highlighterPromise.then((highlighter) => {
-      highlighter.loadLanguage(this.args.language as shiki.BundledLanguage | shiki.LanguageInput | shiki.SpecialLanguage).then(() => {
-        this.asyncHighlightedHTML = highlighter.codeToHtml(this.codeWithoutDiffMarkers, {
-          lang: this.args.language,
-          theme: this.isDarkMode
-            ? (SyntaxHighlightedDiffComponent.highlighterOptionsForDarkMode.themes[0] as string)
-            : (SyntaxHighlightedDiffComponent.highlighterOptionsForLightMode.themes[0] as string),
-          transformers: [transformerNotationDiff()],
-        });
-        this.asyncHighlightedCode = this.args.code;
-      });
+    // Wait for the highlighter promise to load
+    const highlighter = await highlighterPromise;
+
+    // Wait for the language to load
+    await highlighter.loadLanguage(language as shiki.BundledLanguage | shiki.LanguageInput | shiki.SpecialLanguage);
+
+    // Format the code and use it
+    this.asyncHighlightedHTML = highlighter.codeToHtml(codeWithoutDiffMarkers, {
+      lang: language,
+      theme: isDarkMode
+        ? (SyntaxHighlightedDiffComponent.highlighterOptionsForDarkMode.themes[0] as string)
+        : (SyntaxHighlightedDiffComponent.highlighterOptionsForLightMode.themes[0] as string),
+      transformers: [transformerNotationDiff()],
     });
+
+    // Remember which code we've last formatted
+    this.asyncHighlightedCode = code;
+
+    // Remember all parameters used for this task run
+    this.#lastHighlightCodeParams.setParams({
+      isDarkMode: isDarkMode,
+      language: language,
+      codeWithoutDiffMarkers: codeWithoutDiffMarkers,
+    });
+
+    // Ensure we don't run this task too often
+    await timeout(DELAY_BETWEEN_HIGHLIGHT_CODE_TASKS_IN_MS);
```

Comment:
The task code is complex because that's the nature of the task it handles. The code is well structured, variables are well named, comments are thoughtfully placed, and it takes 3 minutes to actually read and understand — nothing hard.